### PR TITLE
[Graph Browser 2] Add flask endpoint for getting statvar hierarchy

### PR DESCRIPTION
**Pull request by chejennifer**

### CODE REVIEW

File: server/routes/api/browser.py [hunk 1]
```diff
@@ -110,3 +110,52 @@ def get_observation_ids():
         dcid = cells[0].get('value', '')
         result[obsDate] = dcid
     return Response(json.dumps(result), 200, mimetype='application/json')
+
+
+def statvar_hierarchy_helper(svg_id, sv_groups, processed_svg, processed_sv,
```
Comment (shifucun) [inline server/routes/api/browser.py:115]: nit: consistently using svg here? example: sv_groups -> svg_map

Comment (shifucun) [reply server/routes/api/browser.py:115]: Also please add a function comment on input and the behavior; Worth to mention that the svg_map is modified to add a "parent" field

File: server/routes/api/browser.py [hunk 2]
```diff
@@ -110,3 +110,52 @@ def get_observation_ids():
...
+        statvar_hierarchy_helper(child_svg_id, sv_groups, processed_svg,
+                                 processed_sv, seen_nodes)
+
+
+@cache.cached(timeout=3600 * 24, query_string=True)  # Cache for one day.
+@bp.route('/statvar-hierarchy/<path:dcid>')
+def get_statvar_hierarchy(dcid):
+    """Returns the stat var groups objects and stat vars objects relevant to a specific dcid.
+    
+    Each stat var group object (keyed by its stat var group id) will have an absolute name, optional list of
```
Comment (shifucun) [inline server/routes/api/browser.py:142]: nit: make line length < 80 chars

File: server/routes/api/browser.py [hunk 3]
```diff
@@ -110,3 +110,52 @@ def get_observation_ids():
...
+    child stat vars, optional list of child stat var groups, and optional list of parent stat var groups.
+
+    Each stat var object (keyed by its stat var id) will have its parent stat var group id.
+    """
+    sv_groups = dc.get_statvar_groups(dcid)
+    processed_svg = {}
+    processed_sv = {}
+    seen_nodes = set()
+    for svg_id in sv_groups.keys():
```
Comment (shifucun) [inline server/routes/api/browser.py:151]: for svg_id, svg in svgs.items():

File: server/routes/api/browser.py [hunk 4]
```diff
@@ -110,3 +110,52 @@ def get_observation_ids():
         dcid = cells[0].get('value', '')
         result[obsDate] = dcid
     return Response(json.dumps(result), 200, mimetype='application/json')
+
+
+def statvar_hierarchy_helper(svg_id, sv_groups, processed_svg, processed_sv,
+                             seen_nodes):
```
Comment (shifucun) [inline server/routes/api/browser.py:116]: seen_nodes -> seen_sv

File: server/routes/api/browser.py [hunk 5]
```diff
@@ -110,3 +110,52 @@ def get_observation_ids():
...
+    sv_group = sv_groups.get(svg_id, {})
+    for child_sv in sv_group.get("childStatVars", []):
+        if child_sv in seen_nodes:
+            continue
+        child_sv_node = {}
```
Comment (shifucun) [inline server/routes/api/browser.py:121]: child_sv_node -> sv2svg; "node" is too general here and does not tell what it does :)

File: server/routes/api/browser.py [hunk 6]
```diff
@@ -110,3 +110,52 @@ def get_observation_ids():
...
+        child_sv_node["parent"] = svg_id
+        processed_sv[child_sv] = child_sv_node
+        seen_nodes.add(child_sv)
+    for child_svg in sv_group.get("childStatVarGroups", []):
+        child_svg_id = child_svg.get("id")
+        child_svg_node = sv_groups[child_svg_id]
```
Comment (shifucun) [inline server/routes/api/browser.py:127]: child_svg_node -> child_svg

File: server/routes/api/browser.py [hunk 7]
```diff
@@ -110,3 +110,52 @@ def get_observation_ids():
...
+        if not "parent" in child_svg_node:
```
Comment (shifucun) [inline server/routes/api/browser.py:128]: it "parent" not in child_svg:

File: server/routes/api/browser.py [hunk 8]
```diff
@@ -110,3 +110,52 @@ def get_observation_ids():
...
+            child_svg_node["parent"] = []
+        child_svg_node["parent"].append(svg_id)
+        processed_svg[child_svg_id] = child_svg_node
```
Comment (shifucun) [inline server/routes/api/browser.py:131]: processed_svg -> processed_svg_map; so this corresponding to the input "svg_map"

File: server/tests/api_browser_test.py [hunk 9]
```diff
@@ -103,3 +103,78 @@ def side_effect(query):
...
+                "childStatVars": ["sv7", "sv8"],
+            }
+        }
+        response = app.test_client().get(
+            'api/browser/statvar-hierarchy/geoId/06')
+        assert response.status_code == 200
+        result = json.loads(response.data)
+        sv_result = result["statVars"]
+        svg_result = result["statVarGroups"]
+        for sv in sv_result.keys():
```
Comment (shifucun) [inline server/tests/api_browser_test.py:174]: instead of testing each individual field, can test the entire result match expected_result?

**Review by shifucun — COMMENTED** [review]
Overall looks good! Some naming suggestions to make the logic easy to follow.

File: server/routes/api/browser.py (hunk 1 above)
Comment (chejennifer) [reply server/routes/api/browser.py]: done

File: server/routes/api/browser.py (hunk 2 above)
Comment (chejennifer) [reply server/routes/api/browser.py]: done

File: server/routes/api/browser.py (hunk 3 above)
Comment (chejennifer) [reply server/routes/api/browser.py]: done

File: server/routes/api/browser.py (hunk 4 above)
Comment (chejennifer) [reply server/routes/api/browser.py]: done

File: server/routes/api/browser.py (hunk 5 above)
Comment (chejennifer) [reply server/routes/api/browser.py]: done

File: server/routes/api/browser.py (hunk 6 above)
Comment (chejennifer) [reply server/routes/api/browser.py]: done

File: server/routes/api/browser.py (hunk 7 above)
Comment (chejennifer) [reply server/routes/api/browser.py]: done

File: server/routes/api/browser.py (hunk 8 above)
Comment (chejennifer) [reply server/routes/api/browser.py]: done

File: server/tests/api_browser_test.py (hunk 9 above)
Comment (chejennifer) [reply server/tests/api_browser_test.py]: done

**Review by shifucun — APPROVED** [review]
thanks for update!

**Pull request merged**